src/Microsoft.AspNetCore.Razor.Language/Legacy/MarkupParser.cs

```diff
@@ -0,0 +1,916 @@
+            if (lastWhitespace != null)
+            {
+                if (Context.DesignTimeMode || !Context.StartOfLine)
+                {
+                    // Markup owns whitespace in design time mode.
+                    Accept(lastWhitespace);
+                    lastWhitespace = null;
+                }
+            }
+
+            PutCurrentBack();
+            if (lastWhitespace != null)
+            {
+                PutBack(lastWhitespace);
+            }
+
+            OtherParserBlock(builder);
+        }
+
+        private void ParseMarkupComment(in SyntaxListBuilder<RazorSyntaxNode> builder)
+        {
+            Assert(SyntaxKind.OpenAngle);
+
+            // Clear the current token builder.
+            builder.Add(OutputAsMarkupLiteral());
+
+            using (var pooledResult = Pool.Allocate<RazorSyntaxNode>())
+            {
+                var htmlCommentBuilder = pooledResult.Builder;
+
+                // Accept the '<', '!' and double-hyphen token at the beginning of the comment block.
+                AcceptAndMoveNext();
+                AcceptAndMoveNext();
+                AcceptAndMoveNext();
+                SpanContext.EditHandler.AcceptedCharacters = AcceptedCharactersInternal.None;
+                htmlCommentBuilder.Add(OutputAsMarkupLiteral());
+
+                SpanContext.EditHandler.AcceptedCharacters = AcceptedCharactersInternal.Whitespace;
+                while (!EndOfFile)
+                {
+                    ParseMarkupNodes(htmlCommentBuilder, ParseMode.Text, t => t.Kind == SyntaxKind.DoubleHyphen);
+                    var lastDoubleHyphen = AcceptAllButLastDoubleHyphens();
+
+                    if (At(SyntaxKind.CloseAngle))
+                    {
+                        // Output the content in the comment block as a separate markup
+                        SpanContext.EditHandler.AcceptedCharacters = AcceptedCharactersInternal.Whitespace;
+                        htmlCommentBuilder.Add(OutputAsMarkupLiteral());
+
+                        // This is the end of a comment block
+                        Accept(lastDoubleHyphen);
+                        AcceptAndMoveNext();
+                        SpanContext.EditHandler.AcceptedCharacters = AcceptedCharactersInternal.None;
+                        htmlCommentBuilder.Add(OutputAsMarkupLiteral());
+                        var commentBlock = SyntaxFactory.MarkupCommentBlock(htmlCommentBuilder.ToList());
+                        builder.Add(commentBlock);
+                        return;
+                    }
+                    else if (lastDoubleHyphen != null)
+                    {
+                        Accept(lastDoubleHyphen);
+                    }
+                }
+
+                builder.Add(OutputAsMarkupLiteral());
+            }
+        }
+
+        private void ParseRazorCommentWithLeadingAndTrailingWhitespace(in SyntaxListBuilder<RazorSyntaxNode> builder)
```

Comment:
This should eventually just be ParseRazorComment yes?

Reply:
`ParseRazorComment` already exists. See line 518.

Reply:
Is ParseRazorComment ever done on its own though?

Reply:
Yes, from CSharpCodeParser